File: validator/graph_loader.py
"""
Загрузчик графов из YAML форматов (DocHub, archlint и callgraph)
"""

import yaml
import networkx as nx
from typing import Dict, Any, List, Union, Tuple


# Graph source type constants
SOURCE_STRUCTURE = 'structure'
SOURCE_BEHAVIOR = 'behavior'


class GraphLoader:
    """Загружает граф из YAML формата (DocHub, archlint или callgraph)"""

    def load_yaml(self, filename: str) -> nx.DiGraph:
        """
        Загружает граф из YAML файла

        Args:
            filename: Путь к YAML файлу

        Returns:
            Направленный граф NetworkX
        """
        with open(filename, 'r', encoding='utf-8') as f:
            data = yaml.safe_load(f)

        return self.parse_yaml(data)

    def load_yaml_with_source(self, filename: str) -> Tuple[nx.DiGraph, str]:
        """
        Загружает граф из YAML файла и возвращает тип источника

        Args:
            filename: Путь к YAML файлу

        Returns:
            Tuple (граф NetworkX, тип источника: 'structure' или 'behavior')
        """
        with open(filename, 'r', encoding='utf-8') as f:
            data = yaml.safe_load(f)

        source = self.detect_source(data)
        graph = self.parse_yaml(data)
        return graph, source

    def detect_source(self, data: Dict[str, Any]) -> str:
        """
        Определяет тип источника: structural (architecture.yaml) или behavioral (callgraph.yaml)

        Callgraph format has 'nodes' and 'edges' with optional 'entry_point'.
        Structural format has 'components' and 'links'.

        Args:
            data: Словарь с данными из YAML

        Returns:
            'behavior' для callgraph, 'structure' для architecture
        """
        if 'nodes' in data and 'edges' in data:
            # Additional check: callgraph nodes have 'package'/'function' fields
            nodes = data.get('nodes', [])
            if isinstance(nodes, list) and nodes:
                first_node = nodes[0] if isinstance(nodes[0], dict) else {}
                if 'package' in first_node or 'function' in first_node or 'entry_point' in data:
                    return SOURCE_BEHAVIOR
        return SOURCE_STRUCTURE

    def parse_yaml(self, data: Dict[str, Any]) -> nx.DiGraph:
        """
        Парсит YAML в граф NetworkX (автоопределение формата)

        Args:
            data: Словарь с данными из YAML

        Returns:
            Направленный граф
        """
        # Check for callgraph format first (nodes/edges with entry_point or function fields)
        if self.detect_source(data) == SOURCE_BEHAVIOR:
            return self._parse_callgraph_format(data)

        components = data.get('components', [])

        # Определяем формат по типу components
        if isinstance(components, list):
            return self._parse_archlint_format(data)
        else:
            return self._parse_dochub_format(data)

    def _parse_archlint_format(self, data: Dict[str, Any]) -> nx.DiGraph:
        """
        Парсит archlint формат:
        components: [{id, title, entity}, ...]
        links: [{from, to, type}, ...]
        """
        graph = nx.DiGraph()

        # Добавляем узлы
        components = data.get('components', [])
        for comp in components:
            if isinstance(comp, dict):
                comp_id = comp.get('id', '')
                if comp_id:
                    graph.add_node(
                        comp_id,
                        title=comp.get('title', ''),
                        entity=comp.get('entity', ''),
                        properties=comp.get('properties', {})
                    )

        # Добавляем рёбра
        links = data.get('links', [])
        for link in links:
            if isinstance(link, dict):
                from_id = link.get('from', '')
                to_id = link.get('to', '')
                if from_id and to_id:
                    graph.add_edge(
                        from_id,
                        to_id,
                        type=link.get('type', ''),
                        method=link.get('method', ''),
                        properties=link.get('properties', {})
                    )

        return graph

    def _parse_callgraph_format(self, data: Dict[str, Any]) -> nx.DiGraph:
        """
        Парсит callgraph формат (из `archlint callgraph`):
        entry_point: "pkg.Type.Method"
        nodes:
          - id: ..., package: ..., function: ..., type: method|function|external
        edges:
          - from: ..., to: ..., call_type: direct|deferred|goroutine
        stats: {...}

        Nodes are mapped to graph nodes preserving function/package metadata.
        Entry point is stored as graph attribute.
        """
        graph = nx.DiGraph()

        entry_point = data.get('entry_point', '')
        if entry_point:
            graph.graph['entry_point'] = entry_point

        # Add nodes
        nodes = data.get('nodes', [])
        for node in nodes:
            if isinstance(node, dict):
                node_id = node.get('id', '')
                if node_id:
                    graph.add_node(
                        node_id,
                        package=node.get('package', ''),
                        function=node.get('function', ''),
                        receiver=node.get('receiver', ''),
                        type=node.get('type', ''),
                        file=node.get('file', ''),
                        line=node.get('line', 0),
                        depth=node.get('depth', 0),
                        # Map to structural field names for validator compatibility
                        title=node.get('function', node_id),
                        entity=node.get('type', ''),
                    )

        # Add edges preserving order
        edges = data.get('edges', [])
        for edge in edges:
            if isinstance(edge, dict):
                from_id = edge.get('from', '')
                to_id = edge.get('to', '')
                if from_id and to_id:
                    # Ensure nodes exist even if not declared in nodes list
                    if from_id not in graph.nodes:
                        graph.add_node(from_id, title=from_id, entity='unknown')
                    if to_id not in graph.nodes:
                        graph.add_node(to_id, title=to_id, entity='unknown')
                    graph.add_edge(
                        from_id,
                        to_id,
                        call_type=edge.get('call_type', 'direct'),
                        line=edge.get('line', 0),
                        # Map to structural field for validator compatibility
                        type=edge.get('call_type', 'direct'),
                    )

        # Store stats as graph metadata
        stats = data.get('stats', {})
        if stats:
            graph.graph['stats'] = stats

        return graph

    def _parse_dochub_format(self, data: Dict[str, Any]) -> nx.DiGraph:
        """
        Парсит DocHub формат:
        components: {id: {title, entity}, ...}
        links: {from_id: [{to, type}, ...], ...}
        """
        graph = nx.DiGraph()

        # Добавляем узлы (components)
        components = data.get('components', {})
        for comp_id, comp_data in components.items():
            if isinstance(comp_data, dict):
                graph.add_node(
                    comp_id,
                    title=comp_data.get('title', ''),
                    entity=comp_data.get('entity', ''),
                    properties=comp_data.get('properties', {})
                )

        # Добавляем рёбра (links)
        links = data.get('links', {})
        for from_id, link_list in links.items():
            if isinstance(link_list, list):
                for link in link_list:
                    to_id = link.get('to')
                    if to_id:
                        graph.add_edge(
                            from_id,
                            to_id,
                            method=link.get('method', ''),
                            type=link.get('type', ''),
                            properties=link.get('properties', {})
                        )

        return graph

File: internal/graphloader/graphloader.go
// Package graphloader загружает архитектурный граф из YAML (3 формата: archlint,
// DocHub, callgraph) — порт validator/graph_loader.py (NetworkX -> Go), направление A.
//
// Это ПРЕДУСЛОВИЕ полного выпила Python: Go-движок (analyzer) парсит только Go-код,
// а УНИВЕРСАЛЬНЫЙ вход (готовый граф из YAML любого формата) грузил Python. Здесь Go
// учится грузить произвольный граф из YAML. Эквивалентность гарантируется golden'ом
// против networkx (тот же YAML -> идентичные узлы/рёбра/атрибуты).
//
// Семантика 1:1 с nx.DiGraph: параллельные (from,to) СХЛОПЫВАЮТСЯ (last-wins по
// атрибутам, как nx.add_edge на существующем ребре); дефолты полей повторяют Python.
package graphloader

import (
	"fmt"
	"os"

	"github.com/mshogin/archlint/internal/model"
	"gopkg.in/yaml.v3"
)

// Тип источника графа (как в Python: structure=architecture, behavior=callgraph).
const (
	SourceStructure = "structure"
	SourceBehavior  = "behavior"
)

// LoadYAML грузит граф из YAML-файла (автоопределение формата).
func LoadYAML(filename string) (*model.Graph, error) {
	g, _, err := LoadYAMLWithSource(filename)

	return g, err
}

// LoadYAMLWithSource грузит граф и возвращает тип источника (structure|behavior).
func LoadYAMLWithSource(filename string) (*model.Graph, string, error) {
	data, err := os.ReadFile(filename) //nolint:gosec // путь — пользовательский вход CLI
	if err != nil {
		return nil, "", fmt.Errorf("read %s: %w", filename, err)
	}

	return ParseYAMLWithSource(data)
}

// ParseYAML парсит YAML-байты в граф (автоопределение формата).
func ParseYAML(data []byte) (*model.Graph, error) {
	g, _, err := ParseYAMLWithSource(data)

	return g, err
}

// ParseYAMLWithSource парсит YAML-байты в граф + тип источника.
func ParseYAMLWithSource(data []byte) (*model.Graph, string, error) {
	var raw map[string]any
	if err := yaml.Unmarshal(data, &raw); err != nil {
		return nil, "", fmt.Errorf("parse yaml: %w", err)
	}

	if raw == nil {
		raw = map[string]any{}
	}

	source := DetectSource(raw)

	return parseData(raw, source), source, nil
}

// DetectSource определяет источник: behavior (callgraph) при наличии nodes+edges и
// признаков callgraph (package/function в первом узле или entry_point); иначе structure.
// 1:1 с Python detect_source (проверка entry_point — только при непустых nodes).
func DetectSource(raw map[string]any) string {
	_, hasNodes := raw["nodes"]
	_, hasEdges := raw["edges"]

	if hasNodes && hasEdges {
		nodes := asSlice(raw["nodes"])
		if len(nodes) > 0 {
			first := asMap(nodes[0])
			_, hasEntry := raw["entry_point"]

			if hasKey(first, "package") || hasKey(first, "function") || hasEntry {
				return SourceBehavior
			}
		}
	}

	return SourceStructure
}

// parseData диспетчеризует по формату (как Python parse_yaml).
func parseData(raw map[string]any, source string) *model.Graph {
	if source == SourceBehavior {
		return parseCallgraph(raw)
	}

	// structure: components-список -> archlint; components-словарь -> DocHub.
	if comp, ok := raw["components"]; ok {
		if _, isMap := comp.(map[string]any); isMap {
			return parseDochub(raw)
		}
	}

	return parseArchlint(raw)
}

// parseArchlint: components:[{id,title,entity,properties}], links:[{from,to,type,method,properties}].
func parseArchlint(raw map[string]any) *model.Graph {
	g := &model.Graph{}
	idx := newEdgeIndex()

	for _, c := range asSlice(raw["components"]) {
		comp := asMap(c)
		id := asString(comp, "id")

		if id == "" {
			continue
		}

		g.Nodes = append(g.Nodes, model.Node{
			ID:     id,
			Title:  asString(comp, "title"),
			Entity: asString(comp, "entity"),
			Attrs: map[string]any{
				"properties": asMapDefault(comp, "properties"),
			},
		})
	}

	for _, l := range asSlice(raw["links"]) {
		link := asMap(l)
		from := asString(link, "from")
		to := asString(link, "to")

		if from == "" || to == "" {
			continue
		}

		idx.set(g, model.Edge{
			From:   from,
			To:     to,
			Type:   asString(link, "type"),
			Method: asString(link, "method"),
			Attrs: map[string]any{
				"properties": asMapDefault(link, "properties"),
			},
		})
	}

	return g
}

// parseCallgraph: entry_point, nodes:[{id,package,function,receiver,type,file,line,depth}],
// edges:[{from,to,call_type,line}], stats. Узлы из рёбер, не объявленные явно, создаются
// (title=id, entity=unknown). 1:1 с Python _parse_callgraph_format.
func parseCallgraph(raw map[string]any) *model.Graph {
	g := &model.Graph{}
	nodeIdx := map[string]int{} // id -> позиция в g.Nodes

	if ep := asString(raw, "entry_point"); ep != "" {
		ensureAttrs(g)
		g.Attrs["entry_point"] = ep
	}

	addNode := func(n model.Node) {
		nodeIdx[n.ID] = len(g.Nodes)
		g.Nodes = append(g.Nodes, n)
	}

	for _, nd := range asSlice(raw["nodes"]) {
		node := asMap(nd)
		id := asString(node, "id")

		if id == "" {
			continue
		}

		typ := asString(node, "type")
		title := asStringDefault(node, "function", id)
		addNode(model.Node{
			ID:     id,
			Title:  title,
			Entity: typ,
			Attrs: map[string]any{
				"package":  asString(node, "package"),
				"function": asString(node, "function"),
				"receiver": asString(node, "receiver"),
				"type":     typ,
				"file":     asString(node, "file"),
				"line":     asInt(node, "line"),
				"depth":    asInt(node, "depth"),
			},
		})
	}

	idx := newEdgeIndex()

	for _, ed := range asSlice(raw["edges"]) {
		edge := asMap(ed)
		from := asString(edge, "from")
		to := asString(edge, "to")

		if from == "" || to == "" {
			continue
		}

		// Узлы из рёбер, не объявленные в nodes -> создаём (как Python).
		for _, end := range []string{from, to} {
			if _, ok := nodeIdx[end]; !ok {
				addNode(model.Node{ID: end, Title: end, Entity: "unknown"})
			}
		}

		callType := asStringDefault(edge, "call_type", "direct")
		idx.set(g, model.Edge{
			From: from,
			To:   to,
			Type: callType,
			Attrs: map[string]any{
				"call_type": callType,
				"line":      asInt(edge, "line"),
			},
		})
	}

	if stats := asMapRaw(raw["stats"]); len(stats) > 0 {
		ensureAttrs(g)
		g.Attrs["stats"] = stats
	}

	return g
}

// parseDochub: components:{id:{title,entity,properties}}, links:{from:[{to,type,method,properties}]}.
func parseDochub(raw map[string]any) *model.Graph {
	g := &model.Graph{}

	comps := asMap(raw["components"])
	for id, cv := range comps {
		comp := asMap(cv)
		if comp == nil {
			continue
		}

		g.Nodes = append(g.Nodes, model.Node{
			ID:     id,
			Title:  asString(comp, "title"),
			Entity: asString(comp, "entity"),
			Attrs: map[string]any{
				"properties": asMapDefault(comp, "properties"),
			},
		})
	}

	idx := newEdgeIndex()

	links := asMap(raw["links"])
	for from, lv := range links {
		for _, l := range asSlice(lv) {
			link := asMap(l)
			to := asString(link, "to")

			if to == "" {
				continue
			}

			idx.set(g, model.Edge{
				From:   from,
				To:     to,
				Type:   asString(link, "type"),
				Method: asString(link, "method"),
				Attrs: map[string]any{
					"properties": asMapDefault(link, "properties"),
				},
			})
		}
	}

	return g
}

// edgeIndex обеспечивает семантику nx.add_edge: одно ребро на (from,to), повторный
// (from,to) ПЕРЕЗАПИСЫВАЕТ атрибуты (last-wins).
type edgeIndex struct {
	pos map[[2]string]int
}

func newEdgeIndex() *edgeIndex { return &edgeIndex{pos: map[[2]string]int{}} }

func (ei *edgeIndex) set(g *model.Graph, e model.Edge) {
	key := [2]string{e.From, e.To}
	if i, ok := ei.pos[key]; ok {
		g.Edges[i] = e // last-wins

		return
	}

	ei.pos[key] = len(g.Edges)
	g.Edges = append(g.Edges, e)
}

func ensureAttrs(g *model.Graph) {
	if g.Attrs == nil {
		g.Attrs = map[string]any{}
	}
}

// --- accessors над map[string]any (yaml.v3) ---

func asString(m map[string]any, key string) string {
	if m == nil {
		return ""
	}

	if v, ok := m[key].(string); ok {
		return v
	}

	return ""
}

func asStringDefault(m map[string]any, key, def string) string {
	if v := asString(m, key); v != "" {
		return v
	}

	return def
}

func asInt(m map[string]any, key string) int {
	if m == nil {
		return 0
	}

	switch v := m[key].(type) {
	case int:
		return v
	case int64:
		return int(v)
	case float64:
		return int(v)
	}

	return 0
}

func asSlice(v any) []any {
	if s, ok := v.([]any); ok {
		return s
	}

	return nil
}

func asMap(v any) map[string]any {
	if m, ok := v.(map[string]any); ok {
		return m
	}

	return nil
}

func asMapRaw(v any) map[string]any { return asMap(v) }

// asMapDefault возвращает вложенный словарь поля или ПУСТОЙ {} (как Python default {}).
func asMapDefault(m map[string]any, key string) map[string]any {
	if m != nil {
		if mm, ok := m[key].(map[string]any); ok {
			return mm
		}
	}

	return map[string]any{}
}

func hasKey(m map[string]any, key string) bool {
	if m == nil {
		return false
	}

	_, ok := m[key]

	return ok
}
